### PR TITLE
LLVM replaces Manual

### PR DESCRIPTION
**Pull request by Rot127**
Hi,

so this is basically a complete rewrite of the generation code. I used LLVM as the source of information and no longer the Programmers Reference Manual. The reason is mainly that the parsing had to be done for each Manual again. This was error prone and could not be tested easily (and there are a lot more reasons).

There are still some non functional things missing in this version (list below), but I won't have that much time in the next two weeks to work on it further, so I just wanted to put it out for you to review.

I am very happy about any remarks, requests or improvement ideas!

Also I really hope the code is easy to understand, maintainable and extendable. If you have any remarks in this regard, please let me know as well!

### Features comparison

| Old                      | New                                | Notes |
|------------------------|----------------------------------|--------------------------|
| v5 - v62 instr set | LLVM instr. set (v5-v68 + HVX)| [n-o-o-n](https://github.com/n-o-o-n/idp_hexagon) added not public instruction to its plugin. We do not have those. |
| Incorrect sub register parsing | Has been fixed || 
| Only GPRs in register profile | All registers in register profile + argument and return register||
| Hard coded calling convention (arg regs/return regs) | CC derived from LLVM | CC for HVX is not added yet. Rizin can not handle more than 10 arg registers. |
| Hard to update (needs parsing code for each Manual ver.) | Easy to update (as long as LLVM does not change too much.)||
| Does not distinguish between different Hexagon versions. | same | This could be problematic because after v66 vector registers were 1024bit wide instead of 512bit. This plugin sets the sizeof vector registers to 1024bit. ||
| No difference between 32bit immediate values and non 32bit values in the syntax. | 32bit prefixed with `##`. Non 32bit with `#` ||
| Neg. immediate values printed as unsigned int | Signed immediate values! | ##0xffffffff -> ##-0x1 - My favorite :D |
| None | Hardware loop support ||
| None | Instr. packet highlighting ||

### TODO

- HVX register (vector registers) have no special type in the register protocol yet. They are treated as GPRs at the moment. See: https://github.com/rizinorg/rizin/issues/1287
- Double registers are listed as independent registers in the register profile. This should be changed.
- Some asm tests fail on purpose (because of the previous points).
- Rizin analysis tests missing -> Tested only by manually checking it in rizin (which works fine).
- ~No UTF-8 free assembly code at the moment (Instr. packets are prefixed with Unicode chars).~ -> UTF-8 chars are removed.
- Missing Control double registers (C21:20 - C29:28) encodings. This is a problem in the LLVM source. Probably have to create a pull request there.
- HVX instructions and register types are not tested yet (didn't get the IDE to compile HVX code in a reasonable time).
- For consistency: All not generated files should be placed in `handwritten` and then moved.
- Set all `rizin` instruction types for suitable instructions in `hexagon_analyses.c`. Not just jmp instructions.
- Adapt syntax highlighting.
- The method for parsing a Duplex and normal instruction are almost the same. This needs some refactoring.
- (some more tiny things. See comments in code.)

### CODE REVIEW

**Comment by XVilka** [comment]
Overall it's a very good approach, I like it. Will check more thoroughly soon

**Comment by XVilka** [comment]
Please add SPDX with license and your copyright as well.

File: rizin/librz/analysis/arch/hexagon/hexagon_analysis.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+
```
Comment (XVilka) [inline rizin/librz/analysis/arch/hexagon/hexagon_analysis.h:4]: Please remove this unnecessary empty line

File: rizin/librz/analysis/arch/hexagon/hexagon_analysis.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+//========================================
+// The following code is generated.
+// Do not edit. Repository of code generator:
+// https://github.com/rizinorg/rz-hexagon
+
+
```
Comment (XVilka) [inline rizin/librz/analysis/arch/hexagon/hexagon_analysis.h:10]: Same here, just one empty line is enough

File: rizin/librz/asm/arch/hexagon/hexagon.c [hunk 3]
```diff
@@ -0,0 +1,722 @@
+// SPDX-FileCopyrightText: 2021 Rot127 <unisono@quyllur.org>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+//========================================
+// The following code is generated.
+// Do not edit. Repository of code generator:
+// https://github.com/rizinorg/rz-hexagon
+
+
```
Comment (XVilka) [inline rizin/librz/asm/arch/hexagon/hexagon.c:9]: And here

File: rizin/librz/asm/arch/hexagon/hexagon.c [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+		case HEX_REG_PRED_REGS_P1:
+			return "P1";
+		case HEX_REG_PRED_REGS_P2:
+			return "P2";
+		case HEX_REG_PRED_REGS_P3:
+			return "P3";
+	}
+}
+
+inline bool hex_if_duplex(uint32_t insn_word)
```
Comment (XVilka) [inline rizin/librz/asm/arch/hexagon/hexagon.c:610]: Please move `{` to the previous line

**Review by XVilka — CHANGES_REQUESTED** [review]
Three more things:
- Please check with `reuse lint` if all files are SPDX-ed properly
- For files that aren't sources, you can use `.reuse/dep5` as in Rizin https://github.com/rizinorg/rizin/blob/dev/.reuse/dep5
- Please copy `.clang-format` from Rizin repo and check if C files are formatted properly https://github.com/rizinorg/rizin/blob/dev/.clang-format

**Review by XVilka — APPROVED** [review]
It's good enough. But having able to operate without UTF-8 is crucial before merging and synchronizing it in Rizin.

**Comment by Rot127** [comment]
Just recognized that two tests fail and I do not know why, yet.
Will make another push today or maybe next week.

Thank you for the feedback so far!

Jul 16, 2021 07:50:04 Anton Kochkov ***@***.***>:

File: .reuse/dep5 [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+Files: binrz/rizin/*
+Copyright: 2009 RadareOrg <pancake@nopcode.org>
+License: LGPL-3.0-only
+
+Files: man/*
+Copyright: 2020 RizinOrg <info@rizin.re>
+License: LGPL-3.0-only
+
+Files: librz/bin/d/*
```
Comment (XVilka) [inline .reuse/dep5:126]: Could you please remove entries unrelated to this particular repo?

File: README.md [hunk 6]
```diff
@@ -1,12 +1,114 @@
...
+  - `llvm_syntax` holds: `$Rdd8 = combine(#0,#$Ii) ; $Rx16 = add($Rx16in,$Rs16)` (the syntax in LLVM style).
+
+    In case of this duplex Instruction it is actually the LLVM-syntax of the high and
+    low instruction concatenated with a semicolon.
+  - `syntax` holds: `Rdd = combine(#0,#Ii) ; Rx = add(Rxin,Rs)`
+  - `Instruction.operands` is a dictionary which contains `Register` and `Immediate` Python objects.
 
-The parsing part based largely on [hexag00n](https://github.com/programa-stic/hexag00n) importer.
 
 # Contributors
```
Comment (XVilka) [inline README.md:114]: Maybe add yourself here ;)

Comment (XVilka) [reply README.md:114]: And obviously at the first place since most of the code is yours now.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (XVilka) [reply README.md:114]: In fact, remove everyone else, except yourself, since you rewrote everything from scratch anyway.

**Review by XVilka — APPROVED** [review]

**Pull request merged**